runtime/argument_validation.go

```diff
@@ -0,0 +1,354 @@
+	case *sema.UInt32Type:
+		_, ok = value.(interpreter.UInt32Value)
+	case *sema.UInt64Type:
+		_, ok = value.(interpreter.UInt64Value)
+	case *sema.UInt128Type:
+		_, ok = value.(interpreter.UInt128Value)
+	case *sema.UInt256Type:
+		_, ok = value.(interpreter.UInt256Value)
+	case *sema.Word8Type:
+		_, ok = value.(interpreter.Word8Value)
+	case *sema.Word16Type:
+		_, ok = value.(interpreter.Word16Value)
+	case *sema.Word32Type:
+		_, ok = value.(interpreter.Word32Value)
+	case *sema.Word64Type:
+		_, ok = value.(interpreter.Word64Value)
+	case *sema.Fix64Type:
+		_, ok = value.(interpreter.Fix64Value)
+	case *sema.UFix64Type:
+		_, ok = value.(interpreter.UFix64Value)
+	case *sema.OptionalType:
+		// Value must be `nil, or must be of type defined in the optional type.
+		if _, ok = value.(interpreter.NilValue); !ok {
+			ok = valueConformsToSemaType(value, typ.Type)
+		}
+	case *sema.AddressType:
+		_, ok = value.(interpreter.AddressValue)
+	case *sema.VariableSizedType:
+		ok = valueConformsToVariableSizedArrayType(value, typ)
+	case *sema.ConstantSizedType:
+		ok = valueConformsToConstantSizedArrayType(value, typ)
+	case *sema.DictionaryType:
+		ok = valueConformsToDictionaryType(value, typ)
+	case *sema.CompositeType:
+		ok = valueConformsToCompositeType(value, typ)
+	case *sema.CapabilityType:
+		_, ok = value.(*interpreter.CapabilityValue)
+
+	// Following types cannot be used as arguments to a script/transaction
+	case *sema.FunctionType:
+		_, ok = value.(interpreter.FunctionValue)
+	case *sema.TransactionType:
+		// false
+	}
+
+	return
+}
+
+func valueConformsToCompositeType(arg interpreter.Value, compositeType *sema.CompositeType) bool {
+	compositeValue, ok := arg.(*interpreter.CompositeValue)
+	if !ok || compositeValue.Kind != compositeType.Kind {
+		return false
+	}
+
+	// Here it is assumed that imported values can only have static fields values,
+	// but not computed field values.
+	if compositeValue.Fields.Len() != len(compositeType.Fields) {
+		return false
+	}
+
+	for _, fieldName := range compositeType.Fields {
+		field, ok := compositeValue.Fields.Get(fieldName)
+		if !ok {
+			return false
+		}
+
+		member, ok := compositeType.Members.Get(fieldName)
+		if !ok {
+			return false
+		}
+
+		if !valueConformsToSemaType(field, member.TypeAnnotation.Type) {
+			return false
+		}
+	}
+
+	return true
+}
+
+func valueConformsToSimpleType(value interpreter.Value, simpleType *sema.SimpleType) (ok bool) {
+	switch simpleType {
+	case sema.StringType:
+		_, ok = value.(*interpreter.StringValue)
+	case sema.BoolType:
+		_, ok = value.(interpreter.BoolValue)
+	case sema.PublicPathType:
+		ok = valueConformsToPathType(value, common.PathDomainPublic)
+	case sema.PrivatePathType:
+		ok = valueConformsToPathType(value, common.PathDomainPrivate)
+	case sema.StoragePathType:
+		ok = valueConformsToPathType(value, common.PathDomainStorage)
+	case sema.PathType, sema.CapabilityPathType:
+		_, ok = value.(interpreter.PathValue)
```

Comment:
for `CapabilityPathType` should check domain is either public or private

Reply:
added with 36327c9